runtime/runtime_plugins.go

```diff
@@ -0,0 +1,216 @@
+	}
+
+	// register before init
+	if params.PluginDir != "" {
+		err = RegisterPluginsFromDir(params.PluginDir)
+		if err != nil {
+			return nil, err
+		}
+	}
+
+	m, plugins, err := initPlugins(params.ID, store, params.ConfigFile)
+	if err != nil {
+		return nil, err
+	}
+
+	var decisionLogger func(context.Context, *server.Info)
+
+	if p, ok := plugins["decision_logs"]; ok {
+		decisionLogger = p.(*logs.Plugin).Log
+
+		if params.DecisionIDFactory == nil {
+			params.DecisionIDFactory = generateDecisionID
+		}
+	}
+
+	rt := &Runtime{
+		Store:          store,
+		Manager:        m,
+		Params:         params,
+		decisionLogger: decisionLogger,
+	}
+
+	return rt, nil
+}
+
+// RegisterBuiltinsFromDir recursively loads all custom builtins into OPA from dir. This function is idempotent.
+func RegisterBuiltinsFromDir(dir string) error {
+	return filepath.Walk(dir, walker("*.builtin.so", registerBuiltinFromFile))
+}
+
+// RegisterPluginsFromDir recursively loads all custom plugins into OPA from dir. This function is idempotent
+func RegisterPluginsFromDir(dir string) error {
+	return filepath.Walk(dir, walker("*.plugin.so", registerPluginFromFile))
+}
+
+// walker returns a walkfunc that performs handler on every file that matches the file glob pattern.
+// it skips all files that do not match pattern TODO: is this what we want?
+func walker(pattern string, handler func(string) error) filepath.WalkFunc {
+	walk := func(path string, f os.FileInfo, err error) error {
+		// if error occurs during traversal to path, exit and crash
+		if err != nil {
+			return err
+		}
+		// skip anything that is a directory
+		if f.IsDir() {
+			return nil
+		}
+
+		if ok, err := filepath.Match(pattern, filepath.Base(path)); err != nil {
+			return err
+		} else if !ok {
+			return nil
+		}
+
+		return handler(path)
+	}
+	return walk
+}
+
+// loads the builtin from a file path
+func registerBuiltinFromFile(path string) error {
+	mod, err := plugin.Open(path)
+	if err != nil {
+		return err
+	}
+	builtinSym, err := mod.Lookup("Builtin")
+	if err != nil {
+		return err
+	}
+	functionSym, err := mod.Lookup("Function")
+	if err != nil {
+		return err
+	}
+
+	// type assert builtin symbol
+	builtin, ok := builtinSym.(*ast.Builtin)
+	if !ok {
+		return fmt.Errorf("symbol Builtin must be of type ast.Builtin")
+	}
+
+	// type assert function symbol
+	switch fnc := functionSym.(type) {
+	case *topdown.BuiltinFunc:
+		ast.RegisterBuiltin(builtin)
+		topdown.RegisterBuiltinFunc(builtin.Name, *fnc)
+	case *topdown.FunctionalBuiltin1:
+		ast.RegisterBuiltin(builtin)
+		topdown.RegisterFunctionalBuiltin1(builtin.Name, *fnc)
+	case *topdown.FunctionalBuiltin2:
+		ast.RegisterBuiltin(builtin)
+		topdown.RegisterFunctionalBuiltin2(builtin.Name, *fnc)
+	case *topdown.FunctionalBuiltin3:
+		ast.RegisterBuiltin(builtin)
+		topdown.RegisterFunctionalBuiltin3(builtin.Name, *fnc)
+	default:
+		return fmt.Errorf("symbol Function was of an unrecognized type")
+	}
+
+	// TODO: replace with logging
```

Comment:
Note: remove before we merge this. Same for below.

Reply:
I think it's useful to have the user know which builtins are being loaded so I'm leaving in the printlns for now.